modin/core/dataframe/base/dataframe/dataframe.py

```diff
@@ -0,0 +1,541 @@
+"""
+
+from abc import ABC, abstractmethod
+from typing import List, Hashable, Optional, Callable, Union, Dict
+from enum import Enum
+
+
+class Axis(Enum):  # noqa: PR01
+    """
+    An enum that represents the axis argument provided to the algebra operators.
+
+    The enum has 3 values - ROW_WISE to represent the row axis, COL_WISE to represent the
+    column axis, and CELL_WISE to represent no axis. ROW_WISE operations iterate over the rows
+    COL_WISE operations over the columns, and CELL_WISE operations over block partitions.
+    """
+
+    ROW_WISE = 0
+    COL_WISE = 1
+    CELL_WISE = None
+
+
+class ModinDataframe(ABC):
+    """
+    An abstract class that represents the Parent class for any Dataframe class.
+
+    This class is intended to specify the behaviors that a Dataframe must implement.
+
+    For more details about how these methods were chosen, please refer to this
+    (http://www.vldb.org/pvldb/vol13/p2033-petersohn.pdf) paper, which specifies
+    a Dataframe algebra that this class exposes.
+    """
+
+    @abstractmethod
+    def mask(
+        self,
+        row_labels: Optional[List[Hashable]] = None,
+        row_positions: Optional[List[int]] = None,
+        col_labels: Optional[List[Hashable]] = None,
+        col_positions: Optional[List[int]] = None,
+    ) -> "ModinDataframe":
+        """
+        Mask rows and columns in the dataframe.
+
+        Allow users to perform selection and projection on the row and column labels (named notation),
+        in addition to the row and column number (positional notation).
+
+        Parameters
+        ----------
+        row_labels : list of hashable, optional
+            The row labels to extract.
+        row_positions : list of int, optional
+            The row positions to extract.
+        col_labels : list of hashable, optional
+            The column labels to extract.
+        col_positions : list of int, optional
+            The column positions to extract.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the mask provided.
+
+        Notes
+        -----
+        If both row_labels and row_positions are provided, an exception is raised.
+        The same rule applies for col_labels and col_positions.
+        """
+        pass
+
+    @abstractmethod
+    def filter_by_types(self, types: List[Hashable]) -> "ModinDataframe":
+        """
+        Allow the user to specify a type or set of types by which to filter the columns.
+
+        Parameters
+        ----------
+        types : list of hashables
+            The types to filter columns by.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe with only the columns whose dtypes appear in types.
+        """
+        pass
+
+    @abstractmethod
+    def map(
+        self,
+        function: Callable,
+        axis: Optional[Union[int, Axis]] = None,
+        dtypes: Optional[str] = None,
```

Comment:
are dtypes str or type? for result_schema the type is Optional[Dict[Hashable, type]] implying that dtypes are of type type.

dtypes: Optional[str] is repeated a few times in this file.

Reply:
We're not a hundred percent sure yet. I think an early proposal was for dtypes to be a dictionary of column names -> types, but some of the algebra operators that have already been implemented have it as a string, with arguments like `copy` - I think the discrepancies are because we have it as a dict in the algebra operators that haven't been implemented because that's the ultimate goal, but as a str for operators that have already been implemented with it as a str to match the implementation.

Reply:
We'll be fixing the `result_schema` argument in the implemented algebra operators in other PRs.